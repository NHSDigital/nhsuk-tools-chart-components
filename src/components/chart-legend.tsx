import React from 'react';
import { Card } from 'nhsuk-react-components';
import './styles/chart-legend.scss';

/**
 * The ChartLegendKey type
 *
 * @param {string} className -  Optional: The classname of the legend key
 * @param {string} wrapperClassName - Optional: The wrapper classname for the card component
 * @param {string} text - The text of the legend key
 * @param {BmiLegendKeys} JSX.Element - Optional: A JSX.Element of the legend key
 */
export type ChartLegendKey = {
  className?: string;
  wrapperClassName?: string;
  text: string;
  element?: JSX.Element;
};

function ChartLegendKeyComponent({
  element,
  text,
  className,
  wrapperClassName,
}: Readonly<ChartLegendKey>): JSX.Element {
  return (
    <div className={`nhsuk-chart-legend-key-wrapper ${wrapperClassName ?? ''}`}>
      <div className={`nhsuk-chart-legend-key ${className ?? ''}`}>
        {element}
      </div>
      {text}
    </div>
  );
}

/**
 * A function to render the chart legend with borders
 * @param {ChartLegendKey[]} legendKey - The legend key item to be rendered
 * @param {stirng} className - The classname for thr components
 */
function ChartLegendBorderedComponent({
  legendKeys,
  className,
}: {
  legendKeys: ChartLegendKey[];
  className: string;
}): JSX.Element {
  return (
    <Card className={`${className} nhsuk-u-padding-3 nhsuk-chart-legend-card`}>
      {legendKeys.map((legendKey) => (
        <ChartLegendKeyComponent key={legendKey.text} {...legendKey} />
      ))}
    </Card>
  );
}

/**
 * A function to render the chart legend without borders
 *
 * @param { ChartLegendKey[]} legendKey - The legend key item to be rendered
 * @param {string} className - The classname for the component
 */
function ChartLegendBorderlessComponent({
  legendKeys,
  className,
}: {
  legendKeys: ChartLegendKey[];
  className: string;
}): JSX.Element {
  return (
    <div className={className}>
      {legendKeys.map((legendKey) => (
        <ChartLegendKeyComponent key={legendKey.text} {...legendKey} />
      ))}
    </div>
  );
}

/**
 * The properties for @ChartLegend component
 *
 * @param {ChartLegendKey[]} legendKeys -  The list of legend keys
 * @param {boolean} borderWrapper -  Optional: A boolean value to enable or disable a border around the legend items with a set max width of 250px
 */
export type ChartLegendProps = {
  legendKeys: ChartLegendKey[];
  borderWrapper?: boolean;
};

/**
 * A component to display a legend associated with a chart, consisting of a card which wraps a set of legend keys.
 *
 * @typedef {object} ChartLegendProps
 * @property {ChartLegendKey[]} legendKeys -  The list of legend keys
 * @property {boolean} borderWrapper - Optional: A boolean value to enable or disable a border around the legend items with a set max width of 250px
 *
 * @param {ChartLegendProps} props
 * @returns {JSX.Element}
 */
export function ChartLegend({
  legendKeys,
  borderWrapper = true,
}: Readonly<ChartLegendProps>): JSX.Element {
  const className =
    'nhsuk-u-margin-top-6 nhsuk-u-margin-bottom-3 nhsuk-u-secondary-text-color nhsuk-body-s';

  return (
    <div aria-hidden="true">
      {borderWrapper ? (
        <ChartLegendBorderedComponent
          legendKeys={legendKeys}
          className={className}
        />
      ) : (
        <ChartLegendBorderlessComponent
          legendKeys={legendKeys}
          className={className}
        />
      )}
    </div>
  );
}
