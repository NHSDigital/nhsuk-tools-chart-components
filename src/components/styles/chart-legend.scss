.nhsuk-chart-legend-card {
  max-width: 250px;
}

.nhsuk-chart-legend-key-wrapper {
  display: flex;
}

.nhsuk-chart-legend-key {
  width: 25px;
  height: 25px;
  margin-right: 10px;
  border-bottom: 1px solid white;
  box-shadow: 0px 2px 0px 0px black;
  margin-bottom: 3px;
}
