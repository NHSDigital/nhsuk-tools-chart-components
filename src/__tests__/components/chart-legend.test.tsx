import { ChartLegend, ChartLegendProps } from '@components/chart-legend';
import { render } from '@testing-library/react';
import React from 'react';

describe('Chart Legend', () => {
  const mockLegendKeys = [
    {
      text: 'Key One',
      wrapperClassName: 'legend-wrapper',
    },
    {
      text: 'Key Two',
      className: 'key-two-class',
    },
    {
      text: 'Key Three',
      element: <span>Key three element</span>,
    },
  ];

  it('renders content as expected', () => {
    const legendProps: ChartLegendProps = {
      legendKeys: mockLegendKeys,
    };

    const { container } = render(<ChartLegend {...legendProps} />);

    expect(container).toMatchSnapshot();
  });

  it(' the chart legend should not contain the card border', () => {
    const legendProps: ChartLegendProps = {
      borderWrapper: false,
      legendKeys: mockLegendKeys,
    };
    const { container } = render(<ChartLegend {...legendProps} />);
    expect(container.querySelector('nhsuk-card')).toBeNull();
    expect(container.querySelector('nhsuk-chart-legend-card')).toBeNull();
  });

  it(' the chart legend should contain the card border', () => {
    const legendProps: ChartLegendProps = {
      borderWrapper: true,
      legendKeys: mockLegendKeys,
    };
    const { container } = render(<ChartLegend {...legendProps} />);
    expect(container.querySelector('nhsuk-card')).toBeDefined();
    expect(container.querySelector('nhsuk-chart-legend-card')).toBeDefined();
  });
});
