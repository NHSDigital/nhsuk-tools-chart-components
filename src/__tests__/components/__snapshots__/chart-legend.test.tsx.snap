// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Legend renders content as expected 1`] = `
<div>
  <div
    aria-hidden="true"
  >
    <div
      class="nhsuk-card nhsuk-u-margin-top-6 nhsuk-u-margin-bottom-3 nhsuk-u-secondary-text-color nhsuk-body-s nhsuk-u-padding-3 nhsuk-chart-legend-card"
    >
      <div
        class="nhsuk-chart-legend-key-wrapper legend-wrapper"
      >
        <div
          class="nhsuk-chart-legend-key "
        />
        Key One
      </div>
      <div
        class="nhsuk-chart-legend-key-wrapper "
      >
        <div
          class="nhsuk-chart-legend-key key-two-class"
        />
        Key Two
      </div>
      <div
        class="nhsuk-chart-legend-key-wrapper "
      >
        <div
          class="nhsuk-chart-legend-key "
        >
          <span>
            Key three element
          </span>
        </div>
        Key Three
      </div>
    </div>
  </div>
</div>
`;
